datasets/flwr_datasets/partitioner/image_semantic_partitioner.py

```diff
@@ -0,0 +1,505 @@
+    gmm_init_params: str
+        The initialization method for the GMM algorithm.
+        Options: ["random", "kmeans", "k-means++"]
+    use_cuda: bool
+        Whether to use CUDA for computation acceleration.
+    shuffle: bool
+        Whether to randomize the order of samples. Shuffling applied after the
+        samples assignment to partitions.
+    sample_seed: int
+        Seed used for sampling.
+    pca_seed: Optional[int]
+        Seed used for PCA dimensionality reduction. If not set, sample_seed is used.
+    gmm_seed: Optional[int]
+        Seed used for GMM clustering. If not set, sample_seed is used.
+    image_column_name: Optional[str]
+        The name of the image column in the dataset. If not set, the first image column
+        is used.
+
+
+    Examples
+    --------
+    >>> from flwr_datasets import FederatedDataset
+    >>> from flwr_datasets.partitioner import ImageSemanticPartitioner
+    >>> partitioner = ImageSemanticPartitioner(
+    >>>     num_partitions=10,
+    >>>     partition_by="label",
+    >>>     pca_components=128,
+    >>>     gmm_max_iter=100,
+    >>>     gmm_init_params="kmeans",
+    >>>     use_cuda=True,
+    >>>     shuffle=True,
+    >>> )
+    >>> fds = FederatedDataset(dataset="mnist", partitioners={"train": partitioner})
+    >>> partition = fds.load_partition(0)
+    >>> print(partition[0])  # Print the first example
+    {'image': <PIL.PngImagePlugin.PngImageFile image mode=L
+    size=28x28 at 0x7FCF49741B10>, 'label': 3}
+    >>> partition_sizes = partition_sizes = [
+    >>>     len(fds.load_partition(partition_id)) for partition_id in range(5)
+    >>> ]
+    >>> print(sorted(partition_sizes))
+    [3163, 5278, 5496, 6320, 9522]
+    """
+
+    def __init__(  # pylint: disable=R0913
+        self,
+        num_partitions: int,
+        partition_by: str,
+        efficient_net_type: int = 3,
+        batch_size: int = 32,
+        pca_components: int = 128,
+        gmm_max_iter: int = 100,
+        gmm_init_params: str = "kmeans",
+        use_cuda: bool = False,
+        shuffle: bool = True,
+        sample_seed: int = 42,
+        pca_seed: Optional[int] = None,
+        gmm_seed: Optional[int] = None,
+        image_column_name: Optional[str] = None,
+    ) -> None:
+        super().__init__()
+        # Attributes based on the constructor
+        self._num_partitions = num_partitions
+        self._partition_by = partition_by
+        self._efficient_net_type = efficient_net_type
+        self._batch_size = batch_size
+        self._efficient_output_size = 1280  # fixed in EfficientNet class
+        self._pca_components = pca_components
+        self._gmm_max_iter = gmm_max_iter
+        self._gmm_init_params = gmm_init_params
+        self._use_cuda = use_cuda
+        self._shuffle = shuffle
+        if pca_seed is None:
+            pca_seed = sample_seed
+        if gmm_seed is None:
+            gmm_seed = sample_seed
```

Comment:
I think if it's None, then it should stay None. These would be a surprising side effect if done as currently. Though, let's add the default values to them in the init.